compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeParamAnalyzer.java

```diff
@@ -580,8 +580,9 @@ private void findTypeParamInTupleForArray(Location loc, BArrayType expType, BTup
         if (actualType.restType != null) {
             tupleTypes.add(actualType.restType);
         }
-        BUnionType tupleElementType = BUnionType.create(null, tupleTypes);
-        findTypeParam(loc, expType.eType, tupleElementType, env, resolvedTypes, result);
+
+        BType type = tupleTypes.size() == 1 ? tupleTypes.iterator().next() : BUnionType.create(null, tupleTypes);
```

Comment:
When the program uses the 'string:char' type, it compiles without any errors. 
```ballerina
import ballerina/io;

public function main() {
    string:Char o = "s";
    var data = ["10", o].map( m => m + 1);
    io:println(data); // prints ["101","s1"]
}
```
I have created a separate [issue](https://github.com/ballerina-platform/ballerina-lang/issues/39700) for this problem

Reply:
This isn't an anonymous function issue - see https://github.com/ballerina-platform/ballerina-lang/issues/39700#issuecomment-1441108180.